meteoloop.1
.TH METEOLOOP "1" "December 2001" "Meteo station tools" Othello
.SH NAME
meteoloop \- compute averages of meteo data
.SH SYNOPSIS
meteoloop [
.B \-d
] [
.B \-F
] [
.BI \-i interval
] [
.BI \-p pidfile
] [
.B \-q
] [
.BI \-r repeats
] [
.B \-n
] [
.BI \-w watchdoginterval
]
.BI \-f conffile
.SH DESCRIPTION
The 
.BR meteoloop (1)
programm connects to the weather station specified in the  configuration
file
.I conffile
and starts reading data from it. It computes averages and extreme values
for all the parameters it finds over a period of one minute, and constructs
an SQL update query to write the information into the database.
Depending on the 
.B \-q
option, the query is either sent directly to the database (as specified
in the configuration file), or to a message queue, where it will later
be picked up by the
.BR meteodequeue (1)
program.
.SH OPTIONS
.TP
.B \-d
increase the debug level by one. Debug messages are written to standard
error.
.TP
.B \-q
send update queries to a message queue, and not to the database
directly. Since the
.I stationdata
table may be locked during a long running query, 
.BR meteoloop (1)
may loose data from the weather station. The use of the message queue
in conjunction with the
.BR meteodequeue (1)
program thus improves reliabilty and performance by decreasing contention
on the 
.I stationdata
table.
.TP
.BI \-f conffile
Use 
.I conffile
to read the configuration, see 
.BR meteo.conf (5)
for the configuration syntax. This argument is mandatory.
.TP
.B \-F
stay in foreground. Useful during debugging.
.TP
.BI \-p pidfile
When running in daemon mode, write the process id of the 
.BR meteoavg (1)
process to this file. Defaults to
.IR /var/run/meteoloop-<station>.pid .
.TP
.BR \-n repeats
Compute only
.I repeats
data points, the default is to run indefinitely.
.TP
.BR \-w watchdoginterval
enable the watchdog timer with an interval of 
.I watchdoginterval
seconds. If no reply is received from the station for an interval of
this duration, the
.BR meteloop (1)
command is restarted with the same arguments.

.SH FILES
.BR meteoloop (1)
writes its process id by default to a file 
.BI /var/run/meteoloop- station .pid

.SH "SEE ALSO"
.BR meteo (1)
.BR meteo.conf (5),
.BR meteoloop (1),
.BR meteograph (1),
.BR meteodequeue (1),
.BR meteocheck (1),
.BR meteojob (1),
.BR meteoavg (1)

.SH AUTHOR
Dr. Andreas Mueller <afm@othello.ch>


NEWS
0.9.17	a new outlet interface can be used to process meteo station data
	directly. In contrast to the map interface, the data is pushed
	to a file or into a pipe by the meteopoll process every minute.

0.9.15	new interface which allows to read current readings from the console
	to display on the web page. The new -m option to meteopoll makes
	sure current data is written to a file (the argument to the option)
	in XML format. Since updates to this file are very frequent, one
	should read the file atomically, one should use the meteomap program
	to read the data.
	fix of a nasty bug in the week label code, causing navigation failure
	at week resolution in some cases

0.9.14	gallery files included as samples
	can now compute sunrise and sunset (through meteosunriset and Sun.cc)

0.9.13	minor bug in meteobrowser.php
	gallery pages

0.9.12	some minor fixes, plus a security fix: remote code execution in
	meteobrowser.php because of missing escapeshellcmd

0.9.11	some bug fixes for GroWeather

0.9.10	turn of CRC0 for GroWeather and possibly other type E weatherlinks.

0.9.9	A hack to protect from IE reading the complete meteobrowser.php
	space offline

0.9.8	fixes a integer overflow bug: to compute the time labels for the
	maps, GraphWindow.cc used the formular (start + end)/2, but since
	about the middle of january, this causes an overflow, so that the
	target ends up in february 1936. Replace that by start/2 + end/2.
	
0.9.7	some bug fixes
	removal of offset configuration variable
	added averages configuration in XML file, so the user now gets full
	control about which averages are compute, and how
	battery and transmitter fields now work
	sunrise and sunset can also be stored in the database (this is a
		waste of space, but who cares about a few gibabytes these
		days)
	added channel of type sun, which indicates the time during which
		the sun is above the horizon

0.9.6	some bug fixes
	removal of header table

0.9.5	some minor fixes

(for further news please check the ChangeLog)

0.9.0	Major rewrite of many aspects of meteo sponsored by the University
	Univates in Lajeado in Rio Grande do Sul, Southern Brazil. Further
	contributions: Jeff Wabik (new derived value computations).

--
$Id: NEWS,v 1.32 2009/01/10 19:00:23 afm Exp $


src/meteoloop.c
/*
 * meteoloop.c 
 *
 * read data from the weather station and update the stationdata table
 *
 * (c) 2001 Dr. Andreas Mueller
 *
 * $Id: meteoloop.c,v 1.9 2002/11/24 23:16:48 afm Exp $
 */
#include <stdlib.h>
#include <stdio.h>
#include <mysql/mysql.h>
#include <meteo.h>
#include <meteodata.h>
#include <sys/types.h>
#include <errno.h>
#include <davis.h>
#include <string.h>
#include <sercom.h>
#include <tcpcom.h>
#include <dloop.h>
#include <dbupdate.h>
#include <unistd.h>
#include <database.h>
#include <msgque.h>
#include <daemon.h>
#include <watchdog.h>
#include <printver.h>
#include <mdebug.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <signal.h>

extern int	optind;
extern char	*optarg;


typedef struct loopdata_s {
	const char	*station;
	const char	*url;
	const char	*queuename;
	int		usequeue;
	int		n;
	meteoconf_t	*mc;
} loopdata_t;

int	minutes;
int	watchinterval = -1;
pid_t	slavepid;

/*
 * the following signal handler ensures that the slave is killed whenever
 * we receive the INT or TERM signal ourselves, 
 */
static void	termhandler(int sig) {
	switch (sig) {
	case SIGTERM:
	case SIGINT:
		if (debug)
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
				"signal %d received, sending to child pid %d",
				sig, slavepid);
		kill(slavepid, sig);
		exit(EXIT_FAILURE);
		break;
	default:
		mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "termhandler called for "
			"wrong signal %d", sig);
		break;
	}
	return;
}

/*
 * childhandler -- this handler is only here so that we can catch the signal
 *                 SIGCHLD, as we are not allowed to ignore it (Posix.1 says)
 *                 but the default action will kill us.
 */
static void	childhandler(int sig) {
	return;
}

/*
 * the setsighandlers function enables or disables signal handling, so
 * that we are always protected from sig chld and handle the TERM and INT
 * signals only if there really is a client to be killed
 */
static void	setsighandlers(int yesno) {
	signal(SIGCHLD, childhandler);
	if (yesno) {
		signal(SIGINT, termhandler);
		signal(SIGTERM, termhandler);
	} else {
		signal(SIGINT, SIG_DFL);
		signal(SIGTERM, SIG_DFL);
	}
}

/*
 * the updateloop function is quite davis specific. It keeps reading meteo
 * data images for n cycles und continuously updates the data in the
 * meteodata_t record. Every time a minute boundary is traversed, the
 * current data is sent to the database using the dbupdate function.
 */
static int	updateloop(dest_t *ddp, loop_t *l, meteodata_t *md, int n,
	const char *station) {
	time_t	now;
	int	i, nn;

	/* if this is a vantage station, wake it up first		*/
	if (l->m->m->flags & COM_VANTAGE)
		vantage_wakeup(l->m->m);

	/* send loop command to station					*/
	put_string(l->m->m, "LOOP");
	if (l->m->m->flags & COM_VANTAGE) {
		put_char(l->m->m, '1');	/* 1 second interval		*/
		nn = n;
		if (n > 9) nn = 9;
		if (n < 1) nn = 1;
		put_char(l->m->m, '0' + nn);
	} else {
		put_unsigned(l->m->m, 0x10000 - n);
	}
	put_char(l->m->m, 0x0d);
	if (debug)
		mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
			"command 'LOOP' command written");

	/* wait for acknowledgment					*/
	if (get_acknowledge(l->m->m) < 0) {
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "no ack received");
		if (watchinterval > 0)
			wd_fire("no ack from station");
		else
			exit(EXIT_FAILURE);
	}

	/* arm the watchdog timer, if set. We do this here so we can	*/
	/* detect a sequence of failed reads. We will reset the watch	*/
	/* on every successfull read					*/
	if (watchinterval > 0)
		wd_arm(watchinterval);

	/* read n reply packets						*/
	for (i = 0; i < n; i++) {

		/* read a new record from the station			*/
		if (debug)
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
				"reading record %d through function %p",
				i, l->read);
		if (l->read(l, md) == 0) {
			if (watchinterval >= 0)
				wd_arm(watchinterval);
		} else {
			mdebug(LOG_ERR, MDEBUG_LOG, 0,
				"retrieving an image failed");
			return -1;
		}
			
		if (debug) {
			meteodata_display(stdout, md);
		}
		time(&now);
		if ((minutes != (now / 60)) && (md->samples > 0)) {
			if (dbupdate(ddp, md, station) < 0) {
				return n - i;
			}
			meteodata_start(md);
			minutes = now / 60;
		}
	}
	/* on an orderly return, disarm the watchdog timer		*/
	wd_disarm();
	return 0;
}

static void	slaveloop(loopdata_t *ld) {
	meteocom_t	*m;
	meteodata_t	*md;
	meteoaccess_t	*d;
	loop_t		*l;
	dest_t		*ddp;
	int		speed = 2400, r;
	struct timeval	tv;
	
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "starting slaveloop with pid %d",
		getpid());

	/* establish the connection to the weather station		*/
	if (0 == strncmp(ld->url, "file://", 7)) {
		speed = xmlconf_get_int(ld->mc, "", "speed", NULL, 0, speed);
		m = sercom_new(ld->mc, (char *)ld->url, speed);
	} else {
		/* setting up a TCP connection can be time consuming...	*/
		if (watchinterval > 0)
			wd_arm(watchinterval);
		m = tcpcom_new(ld->mc, (char *)ld->url);
		wd_disarm();
	}
	if (m == NULL) {
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "cannot open device: %s (%d)",
			  strerror(errno), errno);
		exit(EXIT_FAILURE);
	}

	/* if there are any characters in some buffers, let them drain	*/
	tv.tv_sec = 5; tv.tv_usec = 0;
	com_drain(m, &tv);

	/* create a weather station structure from the connection 	*/
	d = davis_new(m);
	if (d == NULL) {
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "cannot create davis interface");
		exit(EXIT_FAILURE);
	}

	/* create a loop structure for this weather station		*/
	l = dloop_new(d);
	if (l == NULL) {
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "cannot create loop handler");
		exit(EXIT_FAILURE);
	}

	/* establish the database connection				*/
	ddp = dest_new();
	if (ld->usequeue) {
		ddp->type = DEST_MSGQUE;
		ddp->name = strdup(ld->queuename);
		ddp->destdata.msgque = msgque_setup(ld->queuename);
		if (ddp->destdata.msgque < 0) {
			mdebug(LOG_CRIT, MDEBUG_LOG, 0, "cannot open msg queue");
			exit(EXIT_FAILURE);
		}
	} else {
		ddp->type = DEST_MYSQL;
		ddp->name = strdup(xmlconf_get_abs_string(ld->mc,
			"/meteo/database/dbname", "meteo"));
		ddp->destdata.mysql = mc_opendb(ld->mc, O_WRONLY);
		if (NULL == ddp->destdata.mysql) {
			mdebug(LOG_CRIT, MDEBUG_LOG, 0, "failed to connect to "
				"database");
			exit(EXIT_FAILURE);
		}
	}

	/* initialize accumulator data					*/
	md = meteodata_new();
	meteodata_start(md);

	/* initialize the minute counter				*/
	minutes = time(NULL) / 60;

	/* send loop command to station					*/
	while (1) {
		r = updateloop(ddp, l, md, ld->n, ld->station);
		if (debug)
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
				"did %d rounds of updates", ld->n - r);
		if (r != 0) {
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
				"failed with %d remaining updates", r);
			if (watchinterval >= 0)
				wd_fire("updateloop function failed");
			else {
				mdebug(LOG_CRIT, MDEBUG_LOG, 0,
					"loop failed, exiting");
				exit(EXIT_FAILURE);
			}
		}
		wd_disarm();
	}

	/* close the connection to the database				*/
	dest_free(ddp);
}

/*
 * masterloop
 *
 * this function forks a child and returns only if the child has died
 */
static void	masterloop(loopdata_t *l) {
	int	status;

	if (debug)
		mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "starting master loop");

	/* fork a child							*/
	slavepid = fork();
	if (slavepid < 0) {
		mdebug(LOG_CRIT, MDEBUG_LOG, MDEBUG_ERRNO,
			"cannot fork");
		exit(EXIT_FAILURE);
	}

	/* make sure the child is in a new process group. Otherwise	*/
	/* signals generated by the serial line over which the process	*/
	/* is communicating may reach the parent, and kill it, and this	*/
	/* is precisely what we would like to prevent with this master/	*/
	/* slave stuff							*/

	/* the parent waits for the child				*/
	if (slavepid > 0) {
		/* parent sets the pgid of the slave to slave's pid	*/
		setpgid(slavepid, slavepid);

		/* set signal handlers					*/
		setsighandlers(1);

		/* wait for the child					*/
		if (waitpid(slavepid, &status, 0) < 0) {
			/* ignore the ECHILD error, which happens if we	*/
			/* SIG_IGN the CHLD signal, and the child dies	*/
			/* also ignore the EINTR error			*/
			switch (errno) {
			case ECHILD:
			case EINTR:
				break;
			default:
				mdebug(LOG_CRIT, MDEBUG_LOG, MDEBUG_ERRNO,
					"fatal error during wait");
				exit(EXIT_FAILURE);
			}
		}

		/* as the child really should not have exited at all	*/
		/* log an error message					*/
		mdebug(LOG_ERR, MDEBUG_LOG, 0,
			"child %d exited with status %d", slavepid,
			WEXITSTATUS(status));
		if (WIFSIGNALED(status)) {
			mdebug(LOG_ERR, MDEBUG_LOG, 0,
				"exit caused by signal %d",
				WTERMSIG(status));
		}
		setsighandlers(0);

		/* the parent returns as soon as it has waited for the	*/
		/* child						*/
		return;
	}

	/* the child simply calls the slave part			*/
	if (slavepid == 0) {
		/* slave sets its pgid to its own pid			*/
		setpgid(getpid(), getpid());

		/* execute the slave loop				*/
		slaveloop(l);
		exit(EXIT_FAILURE);
	}
}

/*
 * the main function of the meteoloop program simply calls the updateloop
 * function again and again
 */
int	main(int argc, char *argv[]) {
	const char	*url = NULL,
			*station = NULL,
			*queuename = NULL;
	char		*conffilename = METEOCONFFILE;
	loopdata_t	ld;
	int		c, foreground = 0, usemaster = 0;
	char		*pidfilename = "/var/run/meteoloop-%s.pid";

	/* do some initializations that cannot be done in the 		*/
	/* declarations							*/
	ld.usequeue = 0;
	ld.n = 1;

	/* parse command line arguments					*/
	while (EOF != (c = getopt(argc, argv, "dn:Ff:l:p:qw:Vms:")))
		switch (c) {
		case 'l':
			if (mdebug_setup("meteoloop", optarg) < 0) {
				fprintf(stderr, "%s cannot init log\n",
					argv[0]);
				exit(EXIT_FAILURE);
			}
			break;
		case 'd':
			debug++;
			break;
		case 'f':
			conffilename = optarg;
			break;
		case 'F':
			foreground = 1;
			break;
		case 'n':
			ld.n = atoi(optarg);
			if ((ld.n <= 0) || (ld.n > 65535)) {
				mdebug(LOG_CRIT, MDEBUG_LOG, 0,
					"illegal number of repeats: %d", ld.n);
				exit(EXIT_FAILURE);
			}
			break;
		case 'q':
			ld.usequeue = 1;
			if (debug)
				mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
					"updates via msgque requested");
			break;
		case 'p':
			pidfilename = optarg;
			break;
		case 's':
			station = optarg;
			break;
		case 'w':
			watchinterval = atoi(optarg);
			break;
		case 'V':
			fprintver(stdout);
			exit(EXIT_SUCCESS);
			break;
		case 'm':
			usemaster = 1;
			break;
		}

	/* prepare the watchdog timer					*/
	if (watchinterval > 0) {
		if (debug)
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
				"set up the watchdog timer");
		wd_setup(0, NULL);
	}

	/* make sure the station is set					*/
	if (station == NULL) {
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "specify station with -s");
		exit(EXIT_FAILURE);
	}

	/* it is an error not to specify a configuration file		*/
	if (NULL == conffilename) {
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "must specify -f <config>");
		exit(EXIT_FAILURE);
	}
	ld.station = station;

	/* open the configuration file 					*/
	ld.mc = xmlconf_new(conffilename, station);
	if (NULL == ld.mc) {
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "configuration %s invalid", 
			 conffilename);
		exit(EXIT_FAILURE);
	}

	/* extract the information from the configuration data		*/
	ld.url = xmlconf_get_string(ld.mc, "", "url", NULL, 0, url);
	ld.queuename = xmlconf_get_abs_string(ld.mc, "/meteo/database/msgqueue",
		queuename);

	/* become a daemon						*/
	if (!foreground) {
		switch (daemonize(pidfilename, station)) {
		case -1:
			mdebug(LOG_CRIT, MDEBUG_LOG, 0, "failed to daemonize");
			exit(EXIT_FAILURE);
			break;
		case 0:
			/* inside parent				*/
			exit(EXIT_SUCCESS);
			break;
		default:
			break;
		}
	}

	/* start the master/slave stuff (or simply the slave, if the	*/
	/* master is not required					*/
	if (usemaster) {
		while (1) {
			masterloop(&ld);
		}
	} else {
		slaveloop(&ld);
	}

	exit(EXIT_SUCCESS);
}
